extended/src/main/java/io/kubernetes/client/extended/controller/DefaultController.java

```diff
@@ -0,0 +1,192 @@
+package io.kubernetes.client.extended.controller;
+
+import io.kubernetes.client.extended.controller.reconciler.Reconciler;
+import io.kubernetes.client.extended.wait.Wait;
+import io.kubernetes.client.extended.workqueue.RateLimitingQueue;
+import java.time.Duration;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DefaultController implements Controller {
+
+  private static final Logger log = LoggerFactory.getLogger(DefaultController.class);
+
+  private Reconciler reconciler;
+  private String name;
+  private int workerCount;
+  private ExecutorService workerThreadPool;
+
+  private final RateLimitingQueue<Reconciler.Request> workQueue;
+
+  private final Supplier<Boolean>[] readyFuncs;
+  private Duration readyTimeout = Duration.ofSeconds(30);
+  private Duration readyCheckInternal = Duration.ofSeconds(30);
+
+  public DefaultController(
+      Reconciler reconciler,
+      RateLimitingQueue<Reconciler.Request> workQueue,
+      Supplier<Boolean>... readyFuncs) {
+    this.reconciler = reconciler;
+    this.workQueue = workQueue;
+    this.readyFuncs = readyFuncs;
+  }
+
+  @Override
+  public void run() {
```

Comment:
This function is too long/complicated, I'd abstract out an `isSynced()` method or something like that.

Reply:
updated